### PR TITLE
feature(pkg): dune native pinning

### PR DESCRIPTION
**Pull request by rgrinberg**
Introduce a [(package_source ..)] stanza in [dune-project] files to
define dune pins

# TL;DR

A new stanza is available in dune-workspace and dune-project files:
```
(package_source
 (url "...")
 (package (name ..) (version ..))
 (package ..)
 ..)
```

This stanza makes all the package specified in the package_source immediately resolve to `url`.

The work is still WIP, but the important is complete.

### Recursion

`package_source` definitions are recursive (unlike pin-depends). That means dune will scan the `dune-project` files of all dependencies to discover their `package_source` stanzas.

### Compatibility with `pin-depends`

Works transparently with opam pins. If you do `(package_source ..)` on an opam project with a pin, we will respect the pin.

### Subproject Handling

projects are allowed to override the `package_source` assignments of their sub-projects or dependencies. The rules are as follows:
- `a/dune-project` will override any `package_source` set in `a/b/dune-project` b/c `b` is a sub-project of `a`.
- `a/dune-project` can override any `package_source` set in the subprojects that it pulls in `a/dune-project`.

### Workspaces

It's possible to set pins in the workspace file per lock directory

To review this PR, it's best to start by reading the tests.

## TODO

The following elements are still missing:

- proper support for build instructions of dune projects
- recursive project scanning
- workspace sources

## Naming

I dislike the `package_source` name, and I would have much preferred `source`. Unfortunately, `source` is already taken.

## Why a stanza?

I've seen the suggestion of specifying the source inside the `(depends ..)` field floating around more than a few times, so I thought I would explain my rationale for a separate stanza over cramming everything into `(depends ..)`

We could inline the definition of package sources into the `depends` field: 

```
(package
 (depends (a (url git+https://github.com/foo/bar))))
```

But it has a few disadvantages over a separate stanza. Those include:

* When pulling multiple packages from the same url, we don't need to duplicate the URL's
* It's easier to extend to conditional sources with things like `enabled_if` in the stanza. In general, it's easy to extend stanzas with new functionality than an inline specification.
* This syntax might not compose with the full dependency specification from opam
* It's easier to implement because we don't need to touch opam file generation code.

### CODE REVIEW

**Comment by gridbugs** [comment]
I think this merits further discussion. My intuition is that using a stanza will lead to confusion when debugging dependency issues, since the first place people will check is the `depends` stanza and there's no indication there that a package isn't sourced from opam. I think having the spec be easy for humans to read and debug is more important than avoiding duplication and I'm not convinced that the other problems with unlining can't be solved. I'm in the middle of writing a proposal for a new `depends` syntax that should address these issues. I'll link it here when I'm done.

**Comment by gridbugs** [comment]
My proposal for a new `depends` syntax is here: https://github.com/ocaml/dune/issues/9859

**Comment by rgrinberg** [comment]
The way this proposal is implemented, all the details about the frontend are isolated in a single module. This is done on purpose to make it easy to modify such details in the future. Therefore adding inline pins should be rather straightforward in the future. In the meantime, I suggest that we ignore the frontend and focus on the other aspects of the proposal and implementation.

**Comment by Leonidas-from-XIV** [comment]
Regarding `package_source`, which I agree is a bit of a mouthful: How about `override`? Or, `pin`? As you describe, pinning in OPAM is slightly less elaborate, as it doesn't work recursively, but in the basic case is identical and people already know & use the terminology.

**Review by Leonidas-from-XIV — COMMENTED** [review]
Very impressive set of tests, my observations are attached. Looks pretty good overall.

File: test/blackbox-tests/test-cases/pkg/source-stanza/mixed-opam-dune.t [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  > EOF
+  $ cat >_source/bar.opam <<EOF
+  > opam-version: "2.0"
+  > build: [ "echo" "bar" ]
+  > EOF
+
+  $ dune pkg lock
+  Solution for dune.lock:
+  - bar.dev
+  - foo.dev
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/source-stanza/mixed-opam-dune.t:34]: > or set version: X.Y in bar.opam and show that bar.dev instead of bar.X.Y was chosen.

According to the semantics of pins, the version in the `bar.opam` would be ignored in favor of the version set by the caller. So I did your first suggestion.

File: test/blackbox-tests/test-cases/pkg/source-stanza/build-command-dune-project.t [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  >  (depends dune-only mixed template))
+  > EOF
+
+  $ dune pkg lock
+  Solution for dune.lock:
+  - dune-only.dev
+  - mixed.dev
+  - template.dev
+  $ build_command() {
+  > cat dune.lock/$1.pkg | awk '/build/{ p=1 } /^$/{ if (p) {exit} } { if (p) { print $0 } }'
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/source-stanza/build-command-dune-project.t:49]: Yeah, I agree such a tool would be quite useful. We just need to wait for someone to take the plunge on this.

File: src/dune_pkg/mount.mli [hunk 3]
```diff
@@ -0,0 +1,15 @@
+open Import
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_pkg/mount.mli:1]: I couldn't think of something better. Suggestions are welcome.

[the remark this answers is not in the document]
Comment (Leonidas-from-XIV) [reply src/dune_pkg/mount.mli:1]: To me the association with "mounting" things like in a file system made sense.

File: test/blackbox-tests/test-cases/pkg/source-stanza/workspace.t [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  > (package
+  >  (name main)
+  >  (depends foo))
+  > EOF
+
+  $ dune pkg lock
+  File "dune-workspace", line 3, characters 2-6:
+  3 |  (name foo)
+        ^^^^
+  Error: Unknown field name
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/source-stanza/workspace.t:41]: It's not implemented yet. Will come in a subsequent PR.

**Comment by rgrinberg** [comment]
Alright this is good enough for an initial version. Subsequent improvements will come in later versions.

**Comment by rgrinberg** [comment]
@gridbugs @Leonidas-from-XIV this is now ready

File: src/dune_pkg/project_source_stanza.ml [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        Scan_project.eval_url scan_project package.url
+        >>= (function
+         | None -> opam_package package
+         | Some (more_sources, packages) ->
+           dune_package packages package;
+           let more_sources = DB.add_opam_pins more_sources packages in
+           loop stack more_sources))
+  in
+  let+ () = loop [] t in
+  !resolved
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_pkg/project_source_stanza.ml:294]: It's just the simplest. Otherwise we'd need to return the assign packages at every step and then merge them somehow. And the end result would detect errors less promptly without some special data structures.

If you want to try and avoid mutation, feel to try and refactor it later.

File: src/dune_pkg/project_source_stanza.ml [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+      if Path.Source.equal dir dir'
+      then
+        Code_error.raise
+          "two projects in the same directory in the workspace"
+          [ "dir", Path.Source.to_dyn dir ]
+      else if Path.Source.is_descendant dir ~of_:dir'
+      then Ok x'
+      else if Path.Source.is_descendant dir' ~of_:dir
+      then Ok x
+      else Error ()
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_pkg/project_source_stanza.ml:108]: Yes, it's possible. Here's two paths that aren't equal nor are they descendants of each other: `a/b` and `a/c`.

File: src/dune_pkg/project_source_stanza.ml [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+      resolve package.name resolved_package
+  in
+  let rec loop (stack : Local_package.pin list) t =
+    Package_name.Map.values t.map
+    |> Fiber.parallel_iter ~f:(fun (package, _) ->
+      let stack = Stack.push stack package in
+      assign package
+      @@ fun () ->
+      match package.kind with
+      | `Opam -> opam_package package
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_pkg/project_source_stanza.ml:283]: Because opam pins aren't recursive.

File: test/blackbox-tests/test-cases/pkg/source-stanza/unknown-dune-version.t [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  > (lang dune 100.1)
+  > ;; one day we'll get here
+  > EOF
+
+# The location here is messed up b/c we are using a source path (incorrectly)
+# to construct the project
+
+  $ dune pkg lock
+  File "dune-project", line 1, characters 11-16:
+  1 | (lang dune 3.13)
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/source-stanza/unknown-dune-version.t:28]: Yes, unfortunately this is quite difficult to fix. We can only create a dune-project with `Path.Source.t` but this can't be used for correct locations.

File: test/blackbox-tests/test-cases/pkg/source-stanza/override-single-workspace.t [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  > (package_source
+  >  (url "file://$PWD/_bar2")
+  >  (package (name bar)))
+  > EOF
+
+  $ dune pkg lock
+  File "prj2/dune-project", line 4, characters 1-21:
+  4 |  (package (name bar)))
+       ^^^^^^^^^^^^^^^^^^^^
+  Error: package "bar" is defined in more than one source
```
Comment (gridbugs) [inline test/blackbox-tests/test-cases/pkg/source-stanza/override-single-workspace.t:72]: It might be helpful to include both locations in this error message.

**Review by gridbugs — APPROVED** [review]

File: src/dune_pkg/project_source_stanza.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Leonidas-from-XIV) [reply src/dune_pkg/project_source_stanza.ml:110]: Ah, oops, you're right.

**Comment by rgrinberg** [comment]
Alright, I renamed the stanza to `pin`. More fixes will come in subsequent PR's.
